rfcs/recursive-type-unrestriction.md

```diff
@@ -0,0 +1,121 @@
+# Loosening the recursive type restriction
+
+## Summary
+
+Luau supports recursive type aliases, but with an important
+restriction: users can declare functions of recursive types, but *not*
+recursive type functions. This has problems with sophisticated uses of
+types, for example ones which mix recursive types and nested generics.
+This RFC proposes loosening this restriction.
+
+## Motivation
+
+Luau supports recursive type aliases, but with an important restriction:
+users can declare functions of recursive types, such as:
+```lua
+  type Tree<a> = { data: a, children: {Tree<a>} }
+```
+but *not* recursive type functions, such as:
+```lua
+  type TreeWithMap<a> = { ..., map: <b>(a -> b) -> Tree<b> }
+
+```
+These examples cope up naturally in OO code bases with generic types.
+
+## Design
+
+*This section to be filled in once we decide which alternative to use*
+
+## Drawbacks
+
+*This section to be filled in once we decide which alternative to use*
+
+## Alternatives
+
+### Lazy recursive type instantiations
+
+The most permissive change would be to make recursive type
+instantiation lazy rather than strict. In this approach `T<U>` would
+not be instantiated immediately, but only when the body is needed. In
+particular, during unification we can unify `T<U>` with `T<V>` by
+first trying to unify `U` and `V`, and only if that fails try to unify
+the instantiations.
+
+*Advantages*: this allows recursive types with infinite expansions like:
+```lua
+  type Foo<T> = { ..., promises: {Foo<Promise<T>>} }
+```
```

Comment:
may also want to mention the Thunk case:
```lua
type Thunk<T> = (() -> T) | T
```

Reply:
Does that example com from a public repo that we can link to?

Reply:
No public Luau repo, no. If it helps, TypeScript example of part of it is here: https://github.com/reduxjs/redux-thunk/blob/master/src/types.ts

Reply:
Added a link.